### ISSUE TITLE
Bench test on leaf takes down 3-node hub with mirrored stream

### ISSUE THREAD

**Issue by amorey** [nats-io/nats-server#4037]
## Defect

I'm trying to figure out the best way to get data from a set of leaf nodes into a 3-node hub and as part of my tests I've run into an issue with a mirrored stream setup. At a high level, I have a workqueue stream on a leaf node that is mirrored on a 3-node hub using a mirror stream with a replication factor of 3. When I publish messages to the leaf using "nats bench" it works as expected for a while but after a few million messages the hub leader node's memory usage shoots up and eventually the container gets killed by the orchestrator (kubernetes). Then the same thing happens to each new leader node and the cluster never recovers. In the logs I see "high message lag" and lots of "append entries pending" messages. Note that, on occasion I've been able to run the test without this problem and I suspect that it's because the leaf node and the hub leader were in different AZ's so the network connection acted as a throttle to lower the publishing rate.

Here's my current setup:

### Hardware

- gp3 AWS EBS volumes (SSD)
- Each hub node has a 1GB memory limit with env variables `GOMEMLIMIT=980MiB`, `GOGC=25`
- Hub nodes spread across 3 AZs

### Leaf node config

```yaml
# leafnode.yaml
server_name: leafnode
pid_file: "/var/run/nats/nats.pid"
http: 8222
system_account: SYS
debug: false
    
accounts: {
  SYS: {
    users: [
      { user: admin, password: admin }
    ]
  }
}

jetstream {
  store_dir: "/data/jetstream/store"
  max_file_store: 20G
  domain: leaf
}

leafnodes {
  remotes = [
    {
      urls: [
        "nats-leaf://acc:acc@hub.nats-test.svc:7422"
       ]
    }
  ]
}
```

### Leaf stream config

```json
{
  "config": {
    "name": "benchstream",
    "subjects": [
      "TEST.\u003e"
    ],
    "retention": "workqueue",
    "max_consumers": -1,
    "max_msgs_per_subject": -1,
    "max_msgs": -1,
    "max_bytes": -1,
    "max_age": 0,
    "max_msg_size": -1,
    "storage": "file",
    "discard": "old",
    "num_replicas": 1,
    "sealed": false,
    "deny_delete": false,
    "deny_purge": false,
    "allow_rollup_hdrs": false,
    "allow_direct": false,
    "mirror_direct": false
  }
}
```

### Hub node config

As per this [thread](https://github.com/nats-io/nats-server/issues/3724) (https://github.com/nats-io/nats-server/issues/3724) I thought the problem might have to do with the leaf node not having access to flow control message from the hub so I exported the flow control service from the hub. 

```yaml
# hub.yaml (statefulset)
server_name: $POD_NAME
pid_file: "/var/run/nats/nats.pid"
http: 8222
system_account: SYS
debug: false

accounts: {
  SYS: {
    users: [
      { user: admin, password: admin }
    ]
  },
  ACC: {
    jetstream: enabled
    users: [
      { user: acc, password: acc }
    ]
    exports: [
      { service: "$JS.FC.>" }
    ]
  }
}

jetstream {
  store_dir: "/data/jetstream/store"
  max_file_store: 20G
  domain: hub
}
    
cluster {
  name: my-cluster
  port: 6222
  routes [
    nats://hub-0.hub:6222
    nats://hub-1.hub:6222
    nats://hub-2.hub:6222
  ]
  cluster_advertise: $CLUSTER_ADVERTISE
  connect_retries: 30
}

leafnodes {
  port: 7422
}
```

### Hub stream config

```json
{
  "config": {
    "name": "benchstream-remote-mirror",
    "retention": "limits",
    "max_consumers": -1,
    "max_msgs_per_subject": -1,
    "max_msgs": -1,
    "max_bytes": -1,
    "max_age": 0,
    "max_msg_size": -1,
    "storage": "file",
    "discard": "old",
    "num_replicas": 3,
    "mirror": {
      "name": "benchstream",
      "external": {
        "api": "$JS.leaf.API",
        "deliver": ""
      }
    },
    "sealed": false,
    "deny_delete": false,
    "deny_purge": false,
    "allow_rollup_hdrs": false,
    "allow_direct": false,
    "mirror_direct": false
  }
}
```

### Bench command

```sh
nats bench --pub=1 --msgs=20000000 --no-progress TEST.stream_a.1
```

#### Versions of `nats-server` and affected client libraries used:

[25] 2023/04/10 06:43:07.656010 [INF] Starting nats-server
[25] 2023/04/10 06:43:07.656065 [INF]   Version:  2.9.15
[25] 2023/04/10 06:43:07.656070 [INF]   Git:      [b91fa85]
[25] 2023/04/10 06:43:07.656085 [DBG]   Go build: go1.19.6
[25] 2023/04/10 06:43:07.656089 [INF]   Name:     NA5KBJRM67MA4YFWNQ4KGFMO333GACBVEFALAKY54TTXSGA65C23CQKQ
[25] 2023/04/10 06:43:07.656102 [INF]   ID:       NA5KBJRM67MA4YFWNQ4KGFMO333GACBVEFALAKY54TTXSGA65C23CQKQ
[25] 2023/04/10 06:43:07.656129 [DBG] Created system account: "$SYS"
[25] 2023/04/10 06:43:07.657385 [FTL] Error listening on port: 0.0.0.0:4222, "listen tcp 0.0.0.0:4222: bind: address already in use"

#### OS/Container environment:

kubernetes

#### Steps or code to reproduce the issue:

Described above

#### Expected result:

```sh
benchstream: 0
benchstream-remote-mirror: 20,000,000
```

#### Actual result:

Hub outage and

```sh
benchstream: 153,526
benchstream-remote-mirror: 12,453,478
```

Please let me know if I'm doing something wrong or if you have any suggestions on how to pull data from a leaf into a hub in a more robust way!

**Comment by derekcollison** [comment]
You are using normal NATS publish, not jetstream, meaning the bench program is running in pure fire and forget mode.

So first add `-js` to the bench command. If the performance is not what you want we can tweak the async jetstream publish window, etc and you can run more than one publisher as well.

Questions.

1. What server version?
2. Why the setting of GOGC=25? That will make the GC system much more active which can affect performance.

**Comment by amorey** [comment]
Thanks! That did the trick. Sorry, I didn't realize that using `--js` would affect the publishing method.

> What server version?

2.9.15

> Why the setting of GOGC=25? That will make the GC system much more active which can affect performance.

I was playing around with GOMEMLIMIT and GOGC to try to fix the memory issue. Previously I had GOGC set to 100 but I lowered it thinking that doing GC more often would help lower the memory usage. I'll bump it back up.

Some follow-up questions:

1. Is it recommended to have multiple mirrors with non-overlapping subject filters point to a workqueue stream? For example, I want stream_a and stream_b on the hub to mirror the external stream on the leaf with filter subjects "TEST.stream_a.>" and "TEST.stream_b.>" respectively.
2. Do I need to export the `$JS.FC.>` service? If so, can I do it in the default account instead of setting up a new account?
3.  Do I need to export/import any other services or streams in order for flow control and ACK to work properly in a hub/leaf architecture?

**Comment by derekcollison** [comment]
We will be looking at handling the situation better in 2.10 when a NATS core publisher is publishing to a JetStream stream subject.

For the others would need to do a consulting video conference to really dig in to what you are trying to achieve.

**Comment by amorey** [comment]
A consulting video conference would be great. How do I set one up?

**Comment by amorey** [comment]
Now I'm having trouble with a variation on this test. In order to test loss of connectivity, I wrote 20M messages to a workqueue stream on the leaf (using custom code that uses jetstream publish) with the hub disconnected. This completed successfully but when I connected the hub and the remote mirror started mirroring data, hub memory usage shot up and took down the cluster with the same behavior as before. Am I doing something wrong?
